### DISCUSSION TITLE
Keycloak 23.0.7 loosing database connection and healthcheck still OK

### DISCUSSION THREAD

**Discussion by nowheresly** [keycloak/keycloak#28722] [Q&A]
### Description

We encountered the sane issue as https://github.com/keycloak/keycloak/issues/17485

Keycloak healthcheck says everything is fine ( ` "status": "UP"`), but in fact there's no db connection.

Our keycloak is connected to the db through a TCP proxy (HAProxy). For some reason the HAProxy has abruptly been restarted. Yet keycloack is unable to re-create db connection. It remains stuck with no db connection while the healthcheck returns ` "status": "UP"`.

In keycloak logs, the last lines produced are:

`WARN Closing connection in incorrect state VALIDATED`

Manually restarting keycloak fixes the issue. But we are looking for a better solution.

After digging into keycloak / quarkus / agroal documentation and issues, it seems this property:

`quarkus.datasource.jdbc.idle-removal-interval = 10`

could help to force quarkus to re-create db connections (source: https://github.com/quarkusio/quarkus/issues/9123#issuecomment-973237528)

Is there an official way to prevent this stuck scenario?
How could we set this property `quarkus.datasource.jdbc.idle-removal-interval = 10` in keycloak ?

**Comment by jonkoops** [comment]
This is not an enhancement, but a question, moving this to the forum.

**Comment by ihard** [comment]
@Nowheresly, did you manage to find any solution to this problem? I observe similar behavior with loss of connection to the DB in 25.0.4.

**Comment by nowheresly** [comment]
No solution other than restarting keycloak unfortunately

**Comment by shawkins** [comment]
Similar to #33096 and #24493 - the best we can currently do I believe is use quarkus.properties (https://www.keycloak.org/server/configuration) to set connection pool properties that will invalidate the connections. The idle remove property you mention above can be set in quarkus.properties, as can test sql statement which can invalidate the connection when borrowing from the pool.

**Comment by ihard** [comment]
I found a solution without changing the quarkus configuration:
```KC_DB_URL=jdbc:postgresql://master.db-postgres-keycloak-test.service.consul:5432/keycloak?connectTimeout=1&hostRecheckSeconds=2&socketTimeout=600&targetServerType=primary```
